swift/extractor/trap/TrapLabel.h

```diff
@@ -34,6 +34,9 @@ class TrapLabel : public UntypedTrapLabel {
 
   using UntypedTrapLabel::UntypedTrapLabel;
 
+  // we want only TrapArena to create non-default labels
+  friend class TrapArena;
```

Comment:
Why is this necessary? I also do not understand the comment here.

Reply:
we want just one place creating labels, to avoid conflicting labels

Reply:
That doesn't explain why this needs to be a friend. Also with what you just wrote, it more seems like you want some kind of singleton?

Reply:
no, I hate singletons :laughing:

The "one place" was meant in a "one place in code" to make reasoning about it easier when reading the code. We also don't want anyone to create Labels from integers by mistake.

We could have labels as a nested class of whoever needs to create them, and then something like
```
using TrapLabel = WhoeverBuildsLabels::Label;
```

Reply:
Err. Ok, I don't see how you enforce uniqueness of labels then, as `uint64_t id_{0};` in the `TrapArena` is not static.

Reply:
We know there will be one TrapArena (or anything taking its place) per trap file, which is then enough. `TrapLabel` are handled everywhere, so the true risk of conflicting labels is more writing `TrapLabel label{0}` (maybe by someone new reading the public interface of `TrapLabel` and seeing that constructor), rather than multiple `TrapArena`s existing for the same file.

Would you do something else here?

Reply:
Another option is to keep the unsafe "constructor" public (because it might be good to use in tests too) but "name" it something scary: that is, make the constructor itself private, but add a public static function that calls it. Something along the lines of:

```
 private:
  using UntypedTrapLabel::UntypedTrapLabel;
 public:
  // The caller is responsible for ensuring ID uniqueness.
  static TrapLabel UnsafeCreateFromExplicitId(uint64_t id) { return {id}; }
```

Compared to `friend`, this has the benefit of being more restrictive (i.e. it only exposes this specific constructor, but not any other privates), but the downside (or also a benefit?) of losing control over who can call it.

I think `friend` would be fine here, but might suggest moving the `TrapArena` into the same header as `TrapLabel` if they're part of the same API (the user needs to construct the label and then use it - the spelling of those steps is an implementation detail).

Reply:
Right, so that warrants a comment here: https://github.com/github/codeql/pull/9107/files#diff-c25ca6b4acbbb5f8f0d20adf6aea2ceaf6c94f5e7ede65cca4546543669b8e5fR79, not at its current location.

I also still do not understand you this needs to be a friend class.

Reply:
I like the `unsafe` wording, and the option it opens up to create distinct labels in tests